framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -285,104 +283,120 @@ protected boolean areEqualInHierarchy(
+   * of another declared type only if all of the type arguments of the declared type "contain" the
+   * corresponding type arguments of the subtype.
+   *
+   * @param inside a possibly-contained type
+   * @param outside a possibly-containing type
    * @param canBeCovariant whether or not type arguments are allowed to be covariant
    * @return true if inside is contained by outside, or if canBeCovariant == true and {@code inside
    *     <: outside}
    */
   protected boolean isContainedBy(
-      final AnnotatedTypeMirror inside, final AnnotatedTypeMirror outside, boolean canBeCovariant) {
+      AnnotatedTypeMirror inside, AnnotatedTypeMirror outside, boolean canBeCovariant) {
 
-    if (ignoreUninferredTypeArgument(inside) || ignoreUninferredTypeArgument(outside)) {
+    if (shouldIgnoreUninferredTypeArgs(inside) || shouldIgnoreUninferredTypeArgs(outside)) {
       areEqualVisitHistory.put(inside, outside, currentTop, true);
       return true;
     }
 
-    if (outside.getKind() != TypeKind.WILDCARD
-        && !TypesUtils.isCaptured(outside.getUnderlyingType())) {
-      if (canBeCovariant) {
-        return isSubtype(inside, outside, currentTop);
+    if (outside.getKind() == TypeKind.WILDCARD) {
+      Boolean previousResult = areEqualVisitHistory.get(inside, outside, currentTop);
+      if (previousResult != null) {
+        return previousResult;
       }
-      return areEqualInHierarchy(inside, outside);
-    }
+      AnnotatedWildcardType outsideWildcard = (AnnotatedWildcardType) outside;
 
-    AnnotatedTypeMirror outsideUpperBound;
-    AnnotatedTypeMirror outsideLowerBound;
-    if (outside.getKind() == TypeKind.WILDCARD) {
-      outsideUpperBound = ((AnnotatedWildcardType) outside).getExtendsBound();
-      outsideLowerBound = ((AnnotatedWildcardType) outside).getSuperBound();
-    } else if (TypesUtils.isCaptured(outside.getUnderlyingType())) {
-      outsideUpperBound = ((AnnotatedTypeVariable) outside).getUpperBound();
-      outsideLowerBound = ((AnnotatedTypeVariable) outside).getLowerBound();
-    } else {
-      throw new BugInCF("Expected a wildcard or captured type variable, but found " + outside);
+      // Add a placeholder in case of recursion, to prevent infinite regress.
+      areEqualVisitHistory.put(inside, outside, currentTop, true);
+      boolean result =
+          isContainedByBoundType(
+              inside,
+              outsideWildcard.getSuperBound(),
+              outsideWildcard.getExtendsBound(),
+              canBeCovariant);
+      areEqualVisitHistory.put(inside, outside, currentTop, result);
+      return result;
+    } else if ((TypesUtils.isCapturedTypeVariable(outside.getUnderlyingType())
+        && !TypesUtils.isCapturedTypeVariable(inside.getUnderlyingType()))) {
+      // TODO: This branch should be removed after #979 is fixed.
+      // If both outside and inside are captured type variables, they should be equal.
```

Comment:
"they should be equal":  I'm not sure whether this is describing an application invariant that is true whenever control reaches this line, or it is describing a test that will be performed later in the routine.  Could you rewrite to resolve the ambiguity?